// @swagger-api/apidom-ns-asyncapi-3/src/refractor/specification.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
exports.__esModule = true;
exports.default = void 0;
var _apidomNsAsyncapi = require("@swagger-api/apidom-ns-asyncapi-2");
var _index = _interopRequireDefault(require("./visitors/async-api-3/index.cjs"));
var _AsyncApiVersionVisitor = _interopRequireDefault(require("./visitors/async-api-3/AsyncApiVersionVisitor.cjs"));
var _BindingsVisitor = _interopRequireDefault(require("./visitors/async-api-3/channel/BindingsVisitor.cjs"));
var _index2 = _interopRequireDefault(require("./visitors/async-api-3/channel-address-expressions/index.cjs"));
var _index3 = _interopRequireDefault(require("./visitors/async-api-3/channel-bindings/index.cjs"));
var _ServersVisitor = _interopRequireDefault(require("./visitors/async-api-3/channel/ServersVisitor.cjs"));
var _index4 = _interopRequireDefault(require("./visitors/async-api-3/channel/index.cjs"));
var _index5 = _interopRequireDefault(require("./visitors/async-api-3/channels/index.cjs"));
var _ChannelsVisitor = _interopRequireDefault(require("./visitors/async-api-3/components/ChannelsVisitor.cjs"));
var _ExternalDocumentationVisitor = _interopRequireDefault(require("./visitors/async-api-3/components/ExternalDocumentationVisitor.cjs"));
var _OperationsVisitor = _interopRequireDefault(require("./visitors/async-api-3/components/OperationsVisitor.cjs"));
var _RepliesVisitor = _interopRequireDefault(require("./visitors/async-api-3/components/RepliesVisitor.cjs"));
var _ReplyAddressesVisitor = _interopRequireDefault(require("./visitors/async-api-3/components/ReplyAddressesVisitor.cjs"));
var _SchemasVisitor = _interopRequireDefault(require("./visitors/async-api-3/components/SchemasVisitor.cjs"));
var _TagsVisitor = _interopRequireDefault(require("./visitors/async-api-3/components/TagsVisitor.cjs"));
var _index6 = _interopRequireDefault(require("./visitors/async-api-3/components/index.cjs"));
var _index7 = _interopRequireDefault(require("./visitors/async-api-3/contact/index.cjs"));
var _index8 = _interopRequireDefault(require("./visitors/async-api-3/correlation-id/index.cjs"));
var _DefaultContentTypeVisitor = _interopRequireDefault(require("./visitors/async-api-3/DefaultContentTypeVisitor.cjs"));
var _ExternalDocumentationOrReferenceVisitor = _interopRequireDefault(require("./visitors/async-api-3/external-documentation-object/ExternalDocumentationOrReferenceVisitor.cjs"));
var _index9 = _interopRequireDefault(require("./visitors/async-api-3/external-documentation-object/index.cjs"));
var _FallbackVisitor = _interopRequireDefault(require("./visitors/FallbackVisitor.cjs"));
var _IdentifierVisitor = _interopRequireDefault(require("./visitors/async-api-3/IdentifierVisitor.cjs"));
var _info = _interopRequireDefault(require("./visitors/async-api-3/info/info.cjs"));
var _index0 = _interopRequireDefault(require("./visitors/async-api-3/license/index.cjs"));
var _index1 = _interopRequireDefault(require("./visitors/async-api-3/message-bindings/index.cjs"));
var _BindingsVisitor2 = _interopRequireDefault(require("./visitors/async-api-3/message/BindingsVisitor.cjs"));
var _CorrelationIdVisitor = _interopRequireDefault(require("./visitors/async-api-3/message/CorrelationIdVisitor.cjs"));
var _index10 = _interopRequireDefault(require("./visitors/async-api-3/message-example/index.cjs"));
var _ExamplesVisitor = _interopRequireDefault(require("./visitors/async-api-3/message/ExamplesVisitor.cjs"));
var _HeadersVisitor = _interopRequireDefault(require("./visitors/async-api-3/message/HeadersVisitor.cjs"));
var _PayloadVisitor = _interopRequireDefault(require("./visitors/async-api-3/message/PayloadVisitor.cjs"));
var _HeadersVisitor2 = _interopRequireDefault(require("./visitors/async-api-3/message-trait/HeadersVisitor.cjs"));
var _index11 = _interopRequireDefault(require("./visitors/async-api-3/message-trait/index.cjs"));
var _TraitsVisitor = _interopRequireDefault(require("./visitors/async-api-3/message/TraitsVisitor.cjs"));
var _index12 = _interopRequireDefault(require("./visitors/async-api-3/message/index.cjs"));
var _index13 = _interopRequireDefault(require("./visitors/async-api-3/messages/index.cjs"));
var _index14 = _interopRequireDefault(require("./visitors/async-api-3/multiFormatSchema/index.cjs"));
var _index15 = _interopRequireDefault(require("./visitors/async-api-3/oauth-flow/index.cjs"));
var _index16 = _interopRequireDefault(require("./visitors/async-api-3/oauth-flows/index.cjs"));
var _index17 = _interopRequireDefault(require("./visitors/async-api-3/operation-bindings/index.cjs"));
var _BindingsVisitor3 = _interopRequireDefault(require("./visitors/async-api-3/operation/BindingsVisitor.cjs"));
var _ChannelVisitor = _interopRequireDefault(require("./visitors/async-api-3/operation/ChannelVisitor.cjs"));
var _MessagesVisitor = _interopRequireDefault(require("./visitors/async-api-3/operation/MessagesVisitor.cjs"));
var _ReplyVisitor = _interopRequireDefault(require("./visitors/async-api-3/operation/ReplyVisitor.cjs"));
var _index18 = _interopRequireDefault(require("./visitors/async-api-3/operation-reply-address/index.cjs"));
var _AddressVisitor = _interopRequireDefault(require("./visitors/async-api-3/operation-reply/AddressVisitor.cjs"));
var _MessagesVisitor2 = _interopRequireDefault(require("./visitors/async-api-3/operation-reply/MessagesVisitor.cjs"));
var _index19 = _interopRequireDefault(require("./visitors/async-api-3/operation-reply/index.cjs"));
var _SecurityVisitor = _interopRequireDefault(require("./visitors/async-api-3/operation/SecurityVisitor.cjs"));
var _index20 = _interopRequireDefault(require("./visitors/async-api-3/operation-trait/index.cjs"));
var _TraitsVisitor2 = _interopRequireDefault(require("./visitors/async-api-3/operation/TraitsVisitor.cjs"));
var _index21 = _interopRequireDefault(require("./visitors/async-api-3/operation/index.cjs"));
var _index22 = _interopRequireDefault(require("./visitors/async-api-3/operations/index.cjs"));
var _index23 = _interopRequireDefault(require("./visitors/async-api-3/parameter/index.cjs"));
var _index24 = _interopRequireDefault(require("./visitors/async-api-3/parameters/index.cjs"));
var _index25 = _interopRequireDefault(require("./visitors/async-api-3/reference/index.cjs"));
var _SchemaOrReferenceVisitor = _interopRequireDefault(require("./visitors/async-api-3/schema/SchemaOrReferenceVisitor.cjs"));
var _index26 = _interopRequireDefault(require("./visitors/async-api-3/schema/index.cjs"));
var _ScopesVisitor = _interopRequireDefault(require("./visitors/async-api-3/security-scheme/ScopesVisitor.cjs"));
var _index27 = _interopRequireDefault(require("./visitors/async-api-3/security-scheme/index.cjs"));
var _index28 = _interopRequireDefault(require("./visitors/async-api-3/server-bindings/index.cjs"));
var _SecurityVisitor2 = _interopRequireDefault(require("./visitors/async-api-3/server/SecurityVisitor.cjs"));
var _index29 = _interopRequireDefault(require("./visitors/async-api-3/server-variable/index.cjs"));
var _index30 = _interopRequireDefault(require("./visitors/async-api-3/server/index.cjs"));
var _index31 = _interopRequireDefault(require("./visitors/async-api-3/servers/index.cjs"));
var _index32 = _interopRequireDefault(require("./visitors/async-api-3/tag/index.cjs"));
var _index33 = _interopRequireDefault(require("./visitors/async-api-3/tags/index.cjs"));
var _inheritedFixedFields = _interopRequireDefault(require("./visitors/async-api-3/schema/inherited-fixed-fields.cjs"));
var _index34 = _interopRequireDefault(require("./visitors/async-api-3/bindings/amqp/channel-binding/index.cjs"));
var _index35 = _interopRequireDefault(require("./visitors/async-api-3/bindings/amqp/message-binding/index.cjs"));
var _index36 = _interopRequireDefault(require("./visitors/async-api-3/bindings/amqp/operation-binding/index.cjs"));
var _index37 = _interopRequireDefault(require("./visitors/async-api-3/bindings/amqp/server-binding/index.cjs"));
var _index38 = _interopRequireDefault(require("./visitors/async-api-3/bindings/amqp1/channel-binding/index.cjs"));
var _index39 = _interopRequireDefault(require("./visitors/async-api-3/bindings/amqp1/message-binding/index.cjs"));
var _index40 = _interopRequireDefault(require("./visitors/async-api-3/bindings/amqp1/operation-binding/index.cjs"));
var _index41 = _interopRequireDefault(require("./visitors/async-api-3/bindings/amqp1/server-binding/index.cjs"));
var _index42 = _interopRequireDefault(require("./visitors/async-api-3/bindings/anypointmq/channel-binding/index.cjs"));
var _index43 = _interopRequireDefault(require("./visitors/async-api-3/bindings/anypointmq/message-binding/index.cjs"));
var _index44 = _interopRequireDefault(require("./visitors/async-api-3/bindings/anypointmq/operation-binding/index.cjs"));
var _index45 = _interopRequireDefault(require("./visitors/async-api-3/bindings/anypointmq/server-binding/index.cjs"));
var _index46 = _interopRequireDefault(require("./visitors/async-api-3/bindings/http/channel-binding/index.cjs"));
var _index47 = _interopRequireDefault(require("./visitors/async-api-3/bindings/http/message-binding/index.cjs"));
var _index48 = _interopRequireDefault(require("./visitors/async-api-3/bindings/http/operation-binding/index.cjs"));
var _index49 = _interopRequireDefault(require("./visitors/async-api-3/bindings/http/server-binding/index.cjs"));
var _index50 = _interopRequireDefault(require("./visitors/async-api-3/bindings/googlepubsub/channel-binding/index.cjs"));
var _index51 = _interopRequireDefault(require("./visitors/async-api-3/bindings/googlepubsub/message-binding/index.cjs"));
var _index52 = _interopRequireDefault(require("./visitors/async-api-3/bindings/googlepubsub/operation-binding/index.cjs"));
var _index53 = _interopRequireDefault(require("./visitors/async-api-3/bindings/googlepubsub/server-binding/index.cjs"));
var _index54 = _interopRequireDefault(require("./visitors/async-api-3/bindings/ibmmq/channel-binding/index.cjs"));
var _index55 = _interopRequireDefault(require("./visitors/async-api-3/bindings/ibmmq/message-binding/index.cjs"));
var _index56 = _interopRequireDefault(require("./visitors/async-api-3/bindings/ibmmq/operation-binding/index.cjs"));
var _index57 = _interopRequireDefault(require("./visitors/async-api-3/bindings/ibmmq/server-binding/index.cjs"));
var _index58 = _interopRequireDefault(require("./visitors/async-api-3/bindings/jms/channel-binding/index.cjs"));
var _index59 = _interopRequireDefault(require("./visitors/async-api-3/bindings/jms/message-binding/index.cjs"));
var _index60 = _interopRequireDefault(require("./visitors/async-api-3/bindings/jms/operation-binding/index.cjs"));
var _index61 = _interopRequireDefault(require("./visitors/async-api-3/bindings/jms/server-binding/index.cjs"));
var _index62 = _interopRequireDefault(require("./visitors/async-api-3/bindings/kafka/channel-binding/index.cjs"));
var _index63 = _interopRequireDefault(require("./visitors/async-api-3/bindings/kafka/message-binding/index.cjs"));
var _index64 = _interopRequireDefault(require("./visitors/async-api-3/bindings/kafka/operation-binding/index.cjs"));
var _index65 = _interopRequireDefault(require("./visitors/async-api-3/bindings/kafka/server-binding/index.cjs"));
var _index66 = _interopRequireDefault(require("./visitors/async-api-3/bindings/mercure/channel-binding/index.cjs"));
var _index67 = _interopRequireDefault(require("./visitors/async-api-3/bindings/mercure/message-binding/index.cjs"));
var _index68 = _interopRequireDefault(require("./visitors/async-api-3/bindings/mercure/operation-binding/index.cjs"));
var _index69 = _interopRequireDefault(require("./visitors/async-api-3/bindings/mercure/server-binding/index.cjs"));
var _index70 = _interopRequireDefault(require("./visitors/async-api-3/bindings/mqtt/channel-binding/index.cjs"));
var _index71 = _interopRequireDefault(require("./visitors/async-api-3/bindings/mqtt/message-binding/index.cjs"));
var _index72 = _interopRequireDefault(require("./visitors/async-api-3/bindings/mqtt/operation-binding/index.cjs"));
var _index73 = _interopRequireDefault(require("./visitors/async-api-3/bindings/mqtt/server-binding/index.cjs"));
var _index74 = _interopRequireDefault(require("./visitors/async-api-3/bindings/mqtt5/channel-binding/index.cjs"));
var _index75 = _interopRequireDefault(require("./visitors/async-api-3/bindings/mqtt5/message-binding/index.cjs"));
var _index76 = _interopRequireDefault(require("./visitors/async-api-3/bindings/mqtt5/operation-binding/index.cjs"));
var _index77 = _interopRequireDefault(require("./visitors/async-api-3/bindings/mqtt5/server-binding/index.cjs"));
var _index78 = _interopRequireDefault(require("./visitors/async-api-3/bindings/nats/channel-binding/index.cjs"));
var _index79 = _interopRequireDefault(require("./visitors/async-api-3/bindings/nats/message-binding/index.cjs"));
var _index80 = _interopRequireDefault(require("./visitors/async-api-3/bindings/nats/operation-binding/index.cjs"));
var _index81 = _interopRequireDefault(require("./visitors/async-api-3/bindings/nats/server-binding/index.cjs"));
var _index82 = _interopRequireDefault(require("./visitors/async-api-3/bindings/pulsar/channel-binding/index.cjs"));
var _index83 = _interopRequireDefault(require("./visitors/async-api-3/bindings/pulsar/message-binding/index.cjs"));
var _index84 = _interopRequireDefault(require("./visitors/async-api-3/bindings/pulsar/operation-binding/index.cjs"));
var _index85 = _interopRequireDefault(require("./visitors/async-api-3/bindings/pulsar/server-binding/index.cjs"));
var _index86 = _interopRequireDefault(require("./visitors/async-api-3/bindings/redis/channel-binding/index.cjs"));
var _index87 = _interopRequireDefault(require("./visitors/async-api-3/bindings/redis/message-binding/index.cjs"));
var _index88 = _interopRequireDefault(require("./visitors/async-api-3/bindings/redis/operation-binding/index.cjs"));
var _index89 = _interopRequireDefault(require("./visitors/async-api-3/bindings/redis/server-binding/index.cjs"));
var _index90 = _interopRequireDefault(require("./visitors/async-api-3/bindings/sns/channel-binding/index.cjs"));
var _index91 = _interopRequireDefault(require("./visitors/async-api-3/bindings/sns/message-binding/index.cjs"));
var _index92 = _interopRequireDefault(require("./visitors/async-api-3/bindings/sns/operation-binding/index.cjs"));
var _index93 = _interopRequireDefault(require("./visitors/async-api-3/bindings/sns/server-binding/index.cjs"));
var _index94 = _interopRequireDefault(require("./visitors/async-api-3/bindings/solace/channel-binding/index.cjs"));
var _index95 = _interopRequireDefault(require("./visitors/async-api-3/bindings/solace/message-binding/index.cjs"));
var _index96 = _interopRequireDefault(require("./visitors/async-api-3/bindings/solace/operation-binding/index.cjs"));
var _index97 = _interopRequireDefault(require("./visitors/async-api-3/bindings/solace/server-binding/index.cjs"));
var _index98 = _interopRequireDefault(require("./visitors/async-api-3/bindings/sqs/channel-binding/index.cjs"));
var _index99 = _interopRequireDefault(require("./visitors/async-api-3/bindings/sqs/message-binding/index.cjs"));
var _index100 = _interopRequireDefault(require("./visitors/async-api-3/bindings/sqs/operation-binding/index.cjs"));
var _index101 = _interopRequireDefault(require("./visitors/async-api-3/bindings/sqs/server-binding/index.cjs"));
var _index102 = _interopRequireDefault(require("./visitors/async-api-3/bindings/stomp/channel-binding/index.cjs"));
var _index103 = _interopRequireDefault(require("./visitors/async-api-3/bindings/stomp/message-binding/index.cjs"));
var _index104 = _interopRequireDefault(require("./visitors/async-api-3/bindings/stomp/operation-binding/index.cjs"));
var _index105 = _interopRequireDefault(require("./visitors/async-api-3/bindings/stomp/server-binding/index.cjs"));
var _index106 = _interopRequireDefault(require("./visitors/async-api-3/bindings/ws/channel-binding/index.cjs"));
var _index107 = _interopRequireDefault(require("./visitors/async-api-3/bindings/ws/message-binding/index.cjs"));
var _index108 = _interopRequireDefault(require("./visitors/async-api-3/bindings/ws/operation-binding/index.cjs"));
var _index109 = _interopRequireDefault(require("./visitors/async-api-3/bindings/ws/server-binding/index.cjs"));
/**
 * Binding elements.
 */
// AMQP 0-9-1

// AMQP 1.0

// Anypoint MQ

// HTTP

// Google Cloud Pub/Sub

// IBM MQ

// JMS

// Kafka

// Mercure

// MQTT

// MQTT 5

// NATS

// Pulsar

// Redis

// SNS

// Solace

// SQS

// STOMP

// WebSocket

const SchemaSpecification = {
  $visitor: _index26.default,
  fixedFields: {
    ..._inheritedFixedFields.default,
    // validation vocabulary
    // validation keywords for Applying Subschemas With Boolean Logic
    allOf: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Schema.fixedFields.allOf,
    anyOf: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Schema.fixedFields.anyOf,
    oneOf: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Schema.fixedFields.oneOf,
    // validation Keywords for Arrays
    items: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Schema.fixedFields.items,
    // validation Keywords for Objects
    properties: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Schema.fixedFields.properties,
    patternProperties: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Schema.fixedFields.patternProperties,
    dependencies: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Schema.fixedFields.dependencies,
    // validation Vocabulary for Schema Re-Use With "definitions"
    definitions: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Schema.fixedFields.definitions,
    // AsyncAPI vocabulary
    discriminator: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Schema.fixedFields.discriminator,
    externalDocs: _ExternalDocumentationOrReferenceVisitor.default,
    deprecated: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Schema.fixedFields.deprecated
  }
};
const specification = {
  visitors: {
    value: _FallbackVisitor.default,
    document: {
      objects: {
        /**
         * `AsyncApi 3.0.0` specification elements.
         */
        AsyncApi: {
          $visitor: _index.default,
          fixedFields: {
            asyncapi: {
              $ref: '#/visitors/document/objects/AsyncApiVersion'
            },
            id: {
              $ref: '#/visitors/document/objects/Identifier'
            },
            info: {
              $ref: '#/visitors/document/objects/Info'
            },
            servers: {
              $ref: '#/visitors/document/objects/Servers'
            },
            defaultContentType: {
              $ref: '#/visitors/document/objects/DefaultContentType'
            },
            channels: {
              $ref: '#/visitors/document/objects/Channels'
            },
            operations: {
              $ref: '#/visitors/document/objects/Operations'
            },
            components: {
              $ref: '#/visitors/document/objects/Components'
            }
          }
        },
        AsyncApiVersion: {
          $visitor: _AsyncApiVersionVisitor.default
        },
        Identifier: {
          $visitor: _IdentifierVisitor.default
        },
        Info: {
          $visitor: _info.default,
          fixedFields: {
            title: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Info.fixedFields.title,
            version: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Info.fixedFields.version,
            description: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Info.fixedFields.description,
            termsOfService: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Info.fixedFields.termsOfService,
            contact: {
              $ref: '#/visitors/document/objects/Contact'
            },
            license: {
              $ref: '#/visitors/document/objects/License'
            },
            tags: {
              $ref: '#/visitors/document/objects/Tags'
            },
            externalDocs: _ExternalDocumentationOrReferenceVisitor.default
          }
        },
        Contact: {
          $visitor: _index7.default,
          fixedFields: {
            name: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Contact.fixedFields.name,
            url: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Contact.fixedFields.url,
            email: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Contact.fixedFields.email
          }
        },
        License: {
          $visitor: _index0.default,
          fixedFields: {
            name: _apidomNsAsyncapi.specificationObj.visitors.document.objects.License.fixedFields.name,
            url: _apidomNsAsyncapi.specificationObj.visitors.document.objects.License.fixedFields.url
          }
        },
        Servers: {
          $visitor: _index31.default
        },
        Server: {
          $visitor: _index30.default,
          fixedFields: {
            host: {
              $ref: '#/visitors/value'
            },
            protocol: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Server.fixedFields.protocol,
            protocolVersion: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Server.fixedFields.protocolVersion,
            pathname: {
              $ref: '#/visitors/value'
            },
            description: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Server.fixedFields.description,
            title: {
              $ref: '#/visitors/value'
            },
            summary: {
              $ref: '#/visitors/value'
            },
            variables: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Server.fixedFields.variables,
            security: _SecurityVisitor2.default,
            tags: {
              $ref: '#/visitors/document/objects/Tags'
            },
            externalDocs: _ExternalDocumentationOrReferenceVisitor.default,
            bindings: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Server.fixedFields.bindings
          }
        },
        ServerVariable: {
          $visitor: _index29.default,
          fixedFields: {
            enum: _apidomNsAsyncapi.specificationObj.visitors.document.objects.ServerVariable.fixedFields.enum,
            default: _apidomNsAsyncapi.specificationObj.visitors.document.objects.ServerVariable.fixedFields.default,
            description: _apidomNsAsyncapi.specificationObj.visitors.document.objects.ServerVariable.fixedFields.description,
            examples: _apidomNsAsyncapi.specificationObj.visitors.document.objects.ServerVariable.fixedFields.examples
          }
        },
        DefaultContentType: {
          $visitor: _DefaultContentTypeVisitor.default
        },
        Channels: {
          $visitor: _index5.default
        },
        Channel: {
          $visitor: _index4.default,
          fixedFields: {
            address: {
              $ref: '#/visitors/value'
            },
            messages: {
              $ref: '#/visitors/document/objects/Messages'
            },
            title: {
              $ref: '#/visitors/value'
            },
            summary: {
              $ref: '#/visitors/value'
            },
            description: {
              $ref: '#/visitors/value'
            },
            servers: _ServersVisitor.default,
            parameters: {
              $ref: '#/visitors/document/objects/Parameters'
            },
            tags: {
              $ref: '#/visitors/document/objects/Tags'
            },
            externalDocs: _ExternalDocumentationOrReferenceVisitor.default,
            bindings: _BindingsVisitor.default
          }
        },
        ChannelAddressExpressions: {
          $visitor: _index2.default
        },
        Messages: {
          $visitor: _index13.default
        },
        Operations: {
          $visitor: _index22.default
        },
        Operation: {
          $visitor: _index21.default,
          fixedFields: {
            action: {
              $ref: '#/visitors/value'
            },
            channel: _ChannelVisitor.default,
            title: {
              $ref: '#/visitors/value'
            },
            summary: {
              $ref: '#/visitors/value'
            },
            description: {
              $ref: '#/visitors/value'
            },
            security: _SecurityVisitor.default,
            tags: {
              $ref: '#/visitors/document/objects/Tags'
            },
            externalDocs: _ExternalDocumentationOrReferenceVisitor.default,
            bindings: _BindingsVisitor3.default,
            traits: _TraitsVisitor2.default,
            messages: _MessagesVisitor.default,
            reply: _ReplyVisitor.default
          }
        },
        OperationTrait: {
          $visitor: _index20.default,
          fixedFields: {
            title: {
              $ref: '#/visitors/value'
            },
            summary: _apidomNsAsyncapi.specificationObj.visitors.document.objects.OperationTrait.fixedFields.summary,
            description: _apidomNsAsyncapi.specificationObj.visitors.document.objects.OperationTrait.fixedFields.description,
            security: _SecurityVisitor.default,
            tags: {
              $ref: '#/visitors/document/objects/Tags'
            },
            externalDocs: _ExternalDocumentationOrReferenceVisitor.default,
            bindings: _apidomNsAsyncapi.specificationObj.visitors.document.objects.OperationTrait.fixedFields.bindings
          }
        },
        OperationReply: {
          $visitor: _index19.default,
          fixedFields: {
            address: _AddressVisitor.default,
            channel: _ChannelVisitor.default,
            messages: _MessagesVisitor2.default
          }
        },
        OperationReplyAddress: {
          $visitor: _index18.default,
          fixedFields: {
            description: {
              $ref: '#/visitors/value'
            },
            location: {
              $ref: '#/visitors/value'
            }
          }
        },
        Parameters: {
          $visitor: _index24.default
        },
        Parameter: {
          $visitor: _index23.default,
          fixedFields: {
            enum: {
              $ref: '#/visitors/value'
            },
            default: {
              $ref: '#/visitors/value'
            },
            description: {
              $ref: '#/visitors/value'
            },
            examples: {
              $ref: '#/visitors/value'
            },
            location: {
              $ref: '#/visitors/value'
            }
          }
        },
        ServerBindings: {
          $visitor: _index28.default,
          fixedFields: {
            http: {
              $ref: '#/visitors/document/objects/bindings/http/ServerBinding'
            },
            ws: {
              $ref: '#/visitors/document/objects/bindings/ws/ServerBinding'
            },
            kafka: {
              $ref: '#/visitors/document/objects/bindings/kafka/ServerBinding'
            },
            anypointmq: {
              $ref: '#/visitors/document/objects/bindings/anypointmq/ServerBinding'
            },
            amqp: {
              $ref: '#/visitors/document/objects/bindings/amqp/ServerBinding'
            },
            amqp1: {
              $ref: '#/visitors/document/objects/bindings/amqp1/ServerBinding'
            },
            mqtt: {
              $ref: '#/visitors/document/objects/bindings/mqtt/ServerBinding'
            },
            mqtt5: {
              $ref: '#/visitors/document/objects/bindings/mqtt5/ServerBinding'
            },
            nats: {
              $ref: '#/visitors/document/objects/bindings/nats/ServerBinding'
            },
            jms: {
              $ref: '#/visitors/document/objects/bindings/jms/ServerBinding'
            },
            sns: {
              $ref: '#/visitors/document/objects/bindings/sns/ServerBinding'
            },
            solace: {
              $ref: '#/visitors/document/objects/bindings/solace/ServerBinding'
            },
            sqs: {
              $ref: '#/visitors/document/objects/bindings/sqs/ServerBinding'
            },
            stomp: {
              $ref: '#/visitors/document/objects/bindings/stomp/ServerBinding'
            },
            redis: {
              $ref: '#/visitors/document/objects/bindings/redis/ServerBinding'
            },
            mercure: {
              $ref: '#/visitors/document/objects/bindings/mercure/ServerBinding'
            },
            ibmmq: {
              $ref: '#/visitors/document/objects/bindings/ibmmq/ServerBinding'
            },
            googlepubsub: {
              $ref: '#/visitors/document/objects/bindings/googlepubsub/ServerBinding'
            },
            pulsar: {
              $ref: '#/visitors/document/objects/bindings/pulsar/ServerBinding'
            }
          }
        },
        ChannelBindings: {
          $visitor: _index3.default,
          fixedFields: {
            http: {
              $ref: '#/visitors/document/objects/bindings/http/ChannelBinding'
            },
            ws: {
              $ref: '#/visitors/document/objects/bindings/ws/ChannelBinding'
            },
            kafka: {
              $ref: '#/visitors/document/objects/bindings/kafka/ChannelBinding'
            },
            anypointmq: {
              $ref: '#/visitors/document/objects/bindings/anypointmq/ChannelBinding'
            },
            amqp: {
              $ref: '#/visitors/document/objects/bindings/amqp/ChannelBinding'
            },
            amqp1: {
              $ref: '#/visitors/document/objects/bindings/amqp1/ChannelBinding'
            },
            mqtt: {
              $ref: '#/visitors/document/objects/bindings/mqtt/ChannelBinding'
            },
            mqtt5: {
              $ref: '#/visitors/document/objects/bindings/mqtt5/ChannelBinding'
            },
            nats: {
              $ref: '#/visitors/document/objects/bindings/nats/ChannelBinding'
            },
            jms: {
              $ref: '#/visitors/document/objects/bindings/jms/ChannelBinding'
            },
            sns: {
              $ref: '#/visitors/document/objects/bindings/sns/ChannelBinding'
            },
            solace: {
              $ref: '#/visitors/document/objects/bindings/solace/ChannelBinding'
            },
            sqs: {
              $ref: '#/visitors/document/objects/bindings/sqs/ChannelBinding'
            },
            stomp: {
              $ref: '#/visitors/document/objects/bindings/stomp/ChannelBinding'
            },
            redis: {
              $ref: '#/visitors/document/objects/bindings/redis/ChannelBinding'
            },
            mercure: {
              $ref: '#/visitors/document/objects/bindings/mercure/ChannelBinding'
            },
            ibmmq: {
              $ref: '#/visitors/document/objects/bindings/ibmmq/ChannelBinding'
            },
            googlepubsub: {
              $ref: '#/visitors/document/objects/bindings/googlepubsub/ChannelBinding'
            },
            pulsar: {
              $ref: '#/visitors/document/objects/bindings/pulsar/ChannelBinding'
            }
          }
        },
        OperationBindings: {
          $visitor: _index17.default,
          fixedFields: {
            http: {
              $ref: '#/visitors/document/objects/bindings/http/OperationBinding'
            },
            ws: {
              $ref: '#/visitors/document/objects/bindings/ws/OperationBinding'
            },
            kafka: {
              $ref: '#/visitors/document/objects/bindings/kafka/OperationBinding'
            },
            anypointmq: {
              $ref: '#/visitors/document/objects/bindings/anypointmq/OperationBinding'
            },
            amqp: {
              $ref: '#/visitors/document/objects/bindings/amqp/OperationBinding'
            },
            amqp1: {
              $ref: '#/visitors/document/objects/bindings/amqp1/OperationBinding'
            },
            mqtt: {
              $ref: '#/visitors/document/objects/bindings/mqtt/OperationBinding'
            },
            mqtt5: {
              $ref: '#/visitors/document/objects/bindings/mqtt5/OperationBinding'
            },
            nats: {
              $ref: '#/visitors/document/objects/bindings/nats/OperationBinding'
            },
            jms: {
              $ref: '#/visitors/document/objects/bindings/jms/OperationBinding'
            },
            sns: {
              $ref: '#/visitors/document/objects/bindings/sns/OperationBinding'
            },
            solace: {
              $ref: '#/visitors/document/objects/bindings/solace/OperationBinding'
            },
            sqs: {
              $ref: '#/visitors/document/objects/bindings/sqs/OperationBinding'
            },
            stomp: {
              $ref: '#/visitors/document/objects/bindings/stomp/OperationBinding'
            },
            redis: {
              $ref: '#/visitors/document/objects/bindings/redis/OperationBinding'
            },
            mercure: {
              $ref: '#/visitors/document/objects/bindings/mercure/OperationBinding'
            },
            googlepubsub: {
              $ref: '#/visitors/document/objects/bindings/googlepubsub/OperationBinding'
            },
            ibmmq: {
              $ref: '#/visitors/document/objects/bindings/ibmmq/OperationBinding'
            },
            pulsar: {
              $ref: '#/visitors/document/objects/bindings/pulsar/OperationBinding'
            }
          }
        },
        MessageBindings: {
          $visitor: _index1.default,
          fixedFields: {
            http: {
              $ref: '#/visitors/document/objects/bindings/http/MessageBinding'
            },
            ws: {
              $ref: '#/visitors/document/objects/bindings/ws/MessageBinding'
            },
            kafka: {
              $ref: '#/visitors/document/objects/bindings/kafka/MessageBinding'
            },
            anypointmq: {
              $ref: '#/visitors/document/objects/bindings/anypointmq/MessageBinding'
            },
            amqp: {
              $ref: '#/visitors/document/objects/bindings/amqp/MessageBinding'
            },
            amqp1: {
              $ref: '#/visitors/document/objects/bindings/amqp1/MessageBinding'
            },
            mqtt: {
              $ref: '#/visitors/document/objects/bindings/mqtt/MessageBinding'
            },
            mqtt5: {
              $ref: '#/visitors/document/objects/bindings/mqtt5/MessageBinding'
            },
            nats: {
              $ref: '#/visitors/document/objects/bindings/nats/MessageBinding'
            },
            jms: {
              $ref: '#/visitors/document/objects/bindings/jms/MessageBinding'
            },
            sns: {
              $ref: '#/visitors/document/objects/bindings/sns/MessageBinding'
            },
            solace: {
              $ref: '#/visitors/document/objects/bindings/solace/MessageBinding'
            },
            sqs: {
              $ref: '#/visitors/document/objects/bindings/sqs/MessageBinding'
            },
            stomp: {
              $ref: '#/visitors/document/objects/bindings/stomp/MessageBinding'
            },
            redis: {
              $ref: '#/visitors/document/objects/bindings/redis/MessageBinding'
            },
            mercure: {
              $ref: '#/visitors/document/objects/bindings/mercure/MessageBinding'
            },
            ibmmq: {
              $ref: '#/visitors/document/objects/bindings/ibmmq/MessageBinding'
            },
            googlepubsub: {
              $ref: '#/visitors/document/objects/bindings/googlepubsub/MessageBinding'
            },
            pulsar: {
              $ref: '#/visitors/document/objects/bindings/pulsar/MessageBinding'
            }
          }
        },
        Message: {
          $visitor: _index12.default,
          fixedFields: {
            headers: _HeadersVisitor.default,
            payload: _PayloadVisitor.default,
            correlationId: _CorrelationIdVisitor.default,
            contetType: {
              $ref: '#/visitors/value'
            },
            name: {
              $ref: '#/visitors/value'
            },
            title: {
              $ref: '#/visitors/value'
            },
            summary: {
              $ref: '#/visitors/value'
            },
            description: {
              $ref: '#/visitors/value'
            },
            tags: {
              $ref: '#/visitors/document/objects/Tags'
            },
            externalDocs: _ExternalDocumentationOrReferenceVisitor.default,
            bindings: _BindingsVisitor2.default,
            examples: _ExamplesVisitor.default,
            traits: _TraitsVisitor.default
          }
        },
        MessageTrait: {
          $visitor: _index11.default,
          fixedFields: {
            messageId: {
              $ref: '#/visitors/value'
            },
            headers: _HeadersVisitor2.default,
            correlationId: _apidomNsAsyncapi.specificationObj.visitors.document.objects.MessageTrait.fixedFields.correlationId,
            schemaFormat: _apidomNsAsyncapi.specificationObj.visitors.document.objects.MessageTrait.fixedFields.schemaFormat,
            contentType: _apidomNsAsyncapi.specificationObj.visitors.document.objects.MessageTrait.fixedFields.contentType,
            name: _apidomNsAsyncapi.specificationObj.visitors.document.objects.MessageTrait.fixedFields.name,
            title: _apidomNsAsyncapi.specificationObj.visitors.document.objects.MessageTrait.fixedFields.title,
            summary: _apidomNsAsyncapi.specificationObj.visitors.document.objects.MessageTrait.fixedFields.summary,
            description: _apidomNsAsyncapi.specificationObj.visitors.document.objects.MessageTrait.fixedFields.description,
            tags: {
              $ref: '#/visitors/document/objects/Tags'
            },
            externalDocs: _ExternalDocumentationOrReferenceVisitor.default,
            bindings: _apidomNsAsyncapi.specificationObj.visitors.document.objects.MessageTrait.fixedFields.bindings,
            examples: _apidomNsAsyncapi.specificationObj.visitors.document.objects.MessageTrait.fixedFields.examples
          }
        },
        MessageExample: {
          $visitor: _index10.default,
          fixedFields: {
            headers: _apidomNsAsyncapi.specificationObj.visitors.document.objects.MessageExample.fixedFields.headers,
            payload: _apidomNsAsyncapi.specificationObj.visitors.document.objects.MessageExample.fixedFields.payload,
            name: _apidomNsAsyncapi.specificationObj.visitors.document.objects.MessageExample.fixedFields.name,
            summary: _apidomNsAsyncapi.specificationObj.visitors.document.objects.MessageExample.fixedFields.summary
          }
        },
        Tags: {
          $visitor: _index33.default
        },
        Tag: {
          $visitor: _index32.default,
          fixedFields: {
            name: {
              $ref: '#/visitors/value'
            },
            description: {
              $ref: '#/visitors/value'
            },
            externalDocs: _ExternalDocumentationOrReferenceVisitor.default
          }
        },
        ExternalDocumentation: {
          $visitor: _index9.default,
          fixedFields: {
            description: _apidomNsAsyncapi.specificationObj.visitors.document.objects.ExternalDocumentation.fixedFields.description,
            url: _apidomNsAsyncapi.specificationObj.visitors.document.objects.ExternalDocumentation.fixedFields.url
          }
        },
        JSONReference: {
          $ref: '#/visitors/document/objects/Reference'
        },
        Reference: {
          $visitor: _index25.default,
          fixedFields: {
            $ref: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Reference.fixedFields.$ref
          }
        },
        LinkDescription: _apidomNsAsyncapi.specificationObj.visitors.document.objects.LinkDescription,
        Components: {
          $visitor: _index6.default,
          fixedFields: {
            schemas: _SchemasVisitor.default,
            servers: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Components.fixedFields.servers,
            channels: _ChannelsVisitor.default,
            serverVariables: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Components.fixedFields.serverVariables,
            operations: _OperationsVisitor.default,
            messages: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Components.fixedFields.messages,
            securitySchemes: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Components.fixedFields.securitySchemes,
            parameters: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Components.fixedFields.parameters,
            correlationIds: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Components.fixedFields.correlationIds,
            replies: _RepliesVisitor.default,
            replyAddresses: _ReplyAddressesVisitor.default,
            tags: _TagsVisitor.default,
            externalDocs: _ExternalDocumentationVisitor.default,
            operationTraits: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Components.fixedFields.operationTraits,
            messageTraits: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Components.fixedFields.messageTraits,
            serverBindings: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Components.fixedFields.serverBindings,
            channelBindings: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Components.fixedFields.channelBindings,
            operationBindings: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Components.fixedFields.operationBindings,
            messageBindings: _apidomNsAsyncapi.specificationObj.visitors.document.objects.Components.fixedFields.messageBindings
          }
        },
        MultiFormatSchema: {
          $visitor: _index14.default,
          fixedFields: {
            schemaFormat: {
              $ref: '#/visitors/value'
            },
            schema: {
              $ref: '#/visitors/value'
            }
          }
        },
        JSONSchema: SchemaSpecification,
        Schema: SchemaSpecification,
        SecurityScheme: {
          $visitor: _index27.default,
          fixedFields: {
            type: _apidomNsAsyncapi.specificationObj.visitors.document.objects.SecurityScheme.fixedFields.type,
            description: _apidomNsAsyncapi.specificationObj.visitors.document.objects.SecurityScheme.fixedFields.description,
            name: _apidomNsAsyncapi.specificationObj.visitors.document.objects.SecurityScheme.fixedFields.name,
            in: _apidomNsAsyncapi.specificationObj.visitors.document.objects.SecurityScheme.fixedFields.in,
            scheme: _apidomNsAsyncapi.specificationObj.visitors.document.objects.SecurityScheme.fixedFields.scheme,
            bearerFormat: _apidomNsAsyncapi.specificationObj.visitors.document.objects.SecurityScheme.fixedFields.bearerFormat,
            flows: {
              $ref: '#/visitors/document/objects/OAuthFlows'
            },
            openIdConnectUrl: _apidomNsAsyncapi.specificationObj.visitors.document.objects.SecurityScheme.fixedFields.openIdConnectUrl,
            scopes: _ScopesVisitor.default
          }
        },
        OAuthFlows: {
          $visitor: _index16.default,
          fixedFields: {
            implicit: {
              $ref: '#/visitors/document/objects/OAuthFlow'
            },
            password: {
              $ref: '#/visitors/document/objects/OAuthFlow'
            },
            clientCredentials: {
              $ref: '#/visitors/document/objects/OAuthFlow'
            },
            authorizationCode: {
              $ref: '#/visitors/document/objects/OAuthFlow'
            }
          }
        },
        OAuthFlow: {
          $visitor: _index15.default,
          fixedFields: {
            authorizationUrl: _apidomNsAsyncapi.specificationObj.visitors.document.objects.OAuthFlow.fixedFields.authorizationUrl,
            tokenUrl: _apidomNsAsyncapi.specificationObj.visitors.document.objects.OAuthFlow.fixedFields.tokenUrl,
            refreshUrl: _apidomNsAsyncapi.specificationObj.visitors.document.objects.OAuthFlow.fixedFields.refreshUrl,
            availableScopes: _apidomNsAsyncapi.specificationObj.visitors.document.objects.OAuthFlow.fixedFields.scopes
          }
        },
        CorrelationID: {
          $visitor: _index8.default,
          fixedFields: {
            description: _apidomNsAsyncapi.specificationObj.visitors.document.objects.CorrelationID.fixedFields.description,
            location: _apidomNsAsyncapi.specificationObj.visitors.document.objects.CorrelationID.fixedFields.location
          }
        },
        bindings: {
          http: {
            ServerBinding: {
              $visitor: _index49.default
            },
            ChannelBinding: {
              $visitor: _index46.default
            },
            OperationBinding: {
              $visitor: _index48.default,
              fixedFields: {
                type: {
                  $ref: '#/visitors/value'
                },
                method: {
                  $ref: '#/visitors/value'
                },
                query: _SchemaOrReferenceVisitor.default,
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            MessageBinding: {
              $visitor: _index47.default,
              fixedFields: {
                headers: _SchemaOrReferenceVisitor.default,
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            }
          },
          ws: {
            ServerBinding: {
              $visitor: _index109.default
            },
            ChannelBinding: {
              $visitor: _index106.default,
              fixedFields: {
                method: {
                  $ref: '#/visitors/value'
                },
                query: _SchemaOrReferenceVisitor.default,
                headers: _SchemaOrReferenceVisitor.default,
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            OperationBinding: {
              $visitor: _index108.default
            },
            MessageBinding: {
              $visitor: _index107.default
            }
          },
          kafka: {
            ServerBinding: {
              $visitor: _index65.default,
              fixedFields: {
                schemaRegistryUrl: {
                  $ref: '#/visitors/value'
                },
                schemaRegistryVendor: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            ChannelBinding: {
              $visitor: _index62.default,
              fixedFields: {
                topic: {
                  $ref: '#/visitors/value'
                },
                partitions: {
                  $ref: '#/visitors/value'
                },
                replicas: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            OperationBinding: {
              $visitor: _index64.default,
              fixedFields: {
                groupId: _SchemaOrReferenceVisitor.default,
                clientId: _SchemaOrReferenceVisitor.default,
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            MessageBinding: {
              $visitor: _index63.default,
              fixedFields: {
                key: _SchemaOrReferenceVisitor.default,
                schemaIdLocation: {
                  $ref: '#/visitors/value'
                },
                schemaIdPayloadEncoding: {
                  $ref: '#/visitors/value'
                },
                schemaLookupStrategy: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            }
          },
          anypointmq: {
            ServerBinding: {
              $visitor: _index45.default
            },
            ChannelBinding: {
              $visitor: _index42.default,
              fixedFields: {
                destination: {
                  $ref: '#/visitors/value'
                },
                destinationType: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            OperationBinding: {
              $visitor: _index44.default
            },
            MessageBinding: {
              $visitor: _index43.default,
              fixedFields: {
                headers: _SchemaOrReferenceVisitor.default,
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            }
          },
          amqp: {
            ServerBinding: {
              $visitor: _index37.default
            },
            ChannelBinding: {
              $visitor: _index34.default,
              fixedFields: {
                is: {
                  $ref: '#/visitors/value'
                },
                exchange: {
                  $ref: '#/visitors/value'
                },
                queue: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            OperationBinding: {
              $visitor: _index36.default,
              fixedFields: {
                expiration: {
                  $ref: '#/visitors/value'
                },
                userId: {
                  $ref: '#/visitors/value'
                },
                cc: {
                  $ref: '#/visitors/value'
                },
                priority: {
                  $ref: '#/visitors/value'
                },
                deliveryMode: {
                  $ref: '#/visitors/value'
                },
                mandatory: {
                  $ref: '#/visitors/value'
                },
                bcc: {
                  $ref: '#/visitors/value'
                },
                replyTo: {
                  $ref: '#/visitors/value'
                },
                timestamp: {
                  $ref: '#/visitors/value'
                },
                ack: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            MessageBinding: {
              $visitor: _index35.default,
              fixedFields: {
                contentEncoding: {
                  $ref: '#/visitors/value'
                },
                messageType: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            }
          },
          amqp1: {
            ServerBinding: {
              $visitor: _index41.default
            },
            ChannelBinding: {
              $visitor: _index38.default
            },
            OperationBinding: {
              $visitor: _index40.default
            },
            MessageBinding: {
              $visitor: _index39.default
            }
          },
          mqtt: {
            ServerBinding: {
              $visitor: _index73.default,
              fixedFields: {
                clientId: {
                  $ref: '#/visitors/value'
                },
                cleanSession: {
                  $ref: '#/visitors/value'
                },
                lastWill: {
                  $ref: '#/visitors/value'
                },
                keepAlive: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            ChannelBinding: {
              $visitor: _index70.default
            },
            OperationBinding: {
              $visitor: _index72.default,
              fixedFields: {
                qos: {
                  $ref: '#/visitors/value'
                },
                retain: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            MessageBinding: {
              $visitor: _index71.default,
              fixedFields: {
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            }
          },
          mqtt5: {
            ServerBinding: {
              $visitor: _index77.default
            },
            ChannelBinding: {
              $visitor: _index74.default
            },
            OperationBinding: {
              $visitor: _index76.default
            },
            MessageBinding: {
              $visitor: _index75.default
            }
          },
          nats: {
            ServerBinding: {
              $visitor: _index81.default
            },
            ChannelBinding: {
              $visitor: _index78.default
            },
            OperationBinding: {
              $visitor: _index80.default,
              fixedFields: {
                queue: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            MessageBinding: {
              $visitor: _index79.default
            }
          },
          pulsar: {
            ServerBinding: {
              $visitor: _index85.default,
              fixedFields: {
                tenant: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            ChannelBinding: {
              $visitor: _index82.default,
              fixedFields: {
                namespace: {
                  $ref: '#/visitors/value'
                },
                persistence: {
                  $ref: '#/visitors/value'
                },
                compaction: {
                  $ref: '#/visitors/value'
                },
                'geo-replication': {
                  $ref: '#/visitors/value'
                },
                retention: {
                  $ref: '#/visitors/value'
                },
                ttl: {
                  $ref: '#/visitors/value'
                },
                deduplication: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            OperationBinding: {
              $visitor: _index84.default
            },
            MessageBinding: {
              $visitor: _index83.default
            }
          },
          jms: {
            ServerBinding: {
              $visitor: _index61.default
            },
            ChannelBinding: {
              $visitor: _index58.default
            },
            OperationBinding: {
              $visitor: _index60.default
            },
            MessageBinding: {
              $visitor: _index59.default
            }
          },
          sns: {
            ServerBinding: {
              $visitor: _index93.default
            },
            ChannelBinding: {
              $visitor: _index90.default
            },
            OperationBinding: {
              $visitor: _index92.default
            },
            MessageBinding: {
              $visitor: _index91.default
            }
          },
          solace: {
            ServerBinding: {
              $visitor: _index97.default,
              fixedFields: {
                bindingVersion: {
                  $ref: '#/visitors/value'
                },
                msgVpn: {
                  $ref: '#/visitors/value'
                }
              }
            },
            ChannelBinding: {
              $visitor: _index94.default
            },
            OperationBinding: {
              $visitor: _index96.default,
              fixedFields: {
                bindingVersion: {
                  $ref: '#/visitors/value'
                },
                destinations: {
                  $ref: '#/visitors/value'
                }
              }
            },
            MessageBinding: {
              $visitor: _index95.default
            }
          },
          sqs: {
            ServerBinding: {
              $visitor: _index101.default
            },
            ChannelBinding: {
              $visitor: _index98.default
            },
            OperationBinding: {
              $visitor: _index100.default
            },
            MessageBinding: {
              $visitor: _index99.default
            }
          },
          stomp: {
            ServerBinding: {
              $visitor: _index105.default
            },
            ChannelBinding: {
              $visitor: _index102.default
            },
            OperationBinding: {
              $visitor: _index104.default
            },
            MessageBinding: {
              $visitor: _index103.default
            }
          },
          redis: {
            ServerBinding: {
              $visitor: _index89.default
            },
            ChannelBinding: {
              $visitor: _index86.default
            },
            OperationBinding: {
              $visitor: _index88.default
            },
            MessageBinding: {
              $visitor: _index87.default
            }
          },
          mercure: {
            ServerBinding: {
              $visitor: _index69.default
            },
            ChannelBinding: {
              $visitor: _index66.default
            },
            OperationBinding: {
              $visitor: _index68.default
            },
            MessageBinding: {
              $visitor: _index67.default
            }
          },
          googlepubsub: {
            ServerBinding: {
              $visitor: _index53.default
            },
            ChannelBinding: {
              $visitor: _index50.default,
              fixedFields: {
                bindingVersion: {
                  $ref: '#/visitors/value'
                },
                labels: {
                  $ref: '#/visitors/value'
                },
                messageRetentionDuration: {
                  $ref: '#/visitors/value'
                },
                messageStoragePolicy: {
                  $ref: '#/visitors/value'
                },
                schemaSettings: {
                  $ref: '#/visitors/value'
                },
                topic: {
                  $ref: '#/visitors/value'
                }
              }
            },
            OperationBinding: {
              $visitor: _index52.default
            },
            MessageBinding: {
              $visitor: _index51.default,
              fixedFields: {
                bindingVersion: {
                  $ref: '#/visitors/value'
                },
                attributes: {
                  $ref: '#/visitors/value'
                },
                orderingKey: {
                  $ref: '#/visitors/value'
                },
                schema: {
                  $ref: '#/visitors/value'
                }
              }
            }
          },
          ibmmq: {
            ServerBinding: {
              $visitor: _index57.default,
              fixedFields: {
                groupId: {
                  $ref: '#/visitors/value'
                },
                ccdtQueueManagerName: {
                  $ref: '#/visitors/value'
                },
                cipherSpec: {
                  $ref: '#/visitors/value'
                },
                multiEndpointServer: {
                  $ref: '#/visitors/value'
                },
                heartBeatInterval: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            ChannelBinding: {
              $visitor: _index54.default,
              fixedFields: {
                destinationType: {
                  $ref: '#/visitors/value'
                },
                queue: {
                  $ref: '#/visitors/value'
                },
                topic: {
                  $ref: '#/visitors/value'
                },
                maxMsgLength: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            MessageBinding: {
              $visitor: _index55.default,
              fixedFields: {
                type: {
                  $ref: '#/visitors/value'
                },
                headers: {
                  $ref: '#/visitors/value'
                },
                description: {
                  $ref: '#/visitors/value'
                },
                expiry: {
                  $ref: '#/visitors/value'
                },
                bindingVersion: {
                  $ref: '#/visitors/value'
                }
              }
            },
            OperationBinding: {
              $visitor: _index56.default
            }
          }
        }
      }
    }
  }
};
var _default = exports.default = specification;
